### PR TITLE
[DRAFT] Option to do query expansion or not

### LINKED ISSUE #125: Fix GTE-ModernColBERT inference on CPU
As explained in #91, query expansion do not "work" when using flash attention, unattended tokens return embeddings full of zeros.
One issue I did not see coming is that a model trained with FA will ignore those, but then if you do inference on CPU, these tokens are not zeros anymore and aren't trained, which effectively destroys the performance.

One solution is to always train FA models with "attend_to_expansion_tokens" (so those are not zeros even with FA), but a more bulletproof (and fixing existing models such as GTE) is to add a parameter to not do query expansion.

### PR DESCRIPTION
**Pull request by NohTow**
I am adding all the logic to be able not to perform query expansion.
Besides adding the feature in itself, it's needed to fix inference of FA models on CPU because FA on GPU is equivalent to not have query expansion, but then when used on CPU it performs query expansion again and so have bad results, will fix #125.

I'll add more comments before the review.

### CODE REVIEW

**Comment by NohTow** [comment]
> One question, by default following the documentation ModernColBERT and ReasonModernColBERT will yield exact results ? :)

yes, essentially right now ModernColBERT and ReasonModernColBERT when used with FA weren't performing query expansion (implicitely, because the embeddings of padding tokens were zeros). This PR just allows to explicitely code this behavior and so to reproduce the same behavior when being used on CPU.

File: pylate/losses/distillation.py [hunk 1]
```diff
@@ -109,6 +115,7 @@ def forward(
         scores = self.score_metric(
             queries_embeddings,
             documents_embeddings,
+            masks[0] if not do_query_expansion else None,
```
[the remark this answers is not in the document]
Comment (NohTow) [reply pylate/losses/distillation.py:118]: Agreed, think it would be cumbersome

File: pylate/rank/rank.py [hunk 2]
```diff
@@ -102,10 +102,14 @@ def rerank(
...
+            [
+                func_convert_to_tensor(query_document_embeddings)
+                for query_document_embeddings in query_documents_embeddings
+            ],
+            batch_first=True,
+            padding_value=0,
+        )
 
         # Pad the documents embeddings
         query_documents_embeddings = torch.nn.utils.rnn.pad_sequence(
```
[the remark this answers is not in the document]
Comment (NohTow) [reply pylate/rank/rank.py:115]: Ahahah it is actually, I added padding to queries everywhere there was padding to document
The real question is why the original padding (commented as "pad the documents embeddings") is actually padding the queries instead of the documents, as most importantly why the fuck it isn't broken right now

[the remark this answers is not in the document]
Comment (NohTow) [reply pylate/rank/rank.py:115]: Nah ok I'm dumb, we are processing each query iteratively because I did not want to blow the memory up, so we do not have to pad the queries
And the comment is actually correct because it's the documents embeddings _respective to a give query_
